### ISSUE TITLE
Decimal to double

### ISSUE THREAD

**Issue by scherepanov** [fastfloat/fast_float#133]
Hi,
I am writing conversion float -> double, with requirement to_string(float) = to_string(double)

I am using Dragonbox to convert float to chars, and fast_float to convert chars to double.

double floatToDouble(float f) {
  constexpr int buffer_length = 1 + // for '\0'
      jkj::dragonbox::max_output_string_length<jkj::dragonbox::ieee754_binary64>;
  char buffer[buffer_length];
  char* end_ptr = jkj::dragonbox::to_chars(f, buffer);
  double d;
  fast_float::from_chars(buffer, end_ptr, d);
  return d;
}

Works perfect! Both Dragonbox and fast_float are very fast!!! Big thanks for nice library!

At same time, intermediate chars format sounds excessive.

Dragonbox offer conversion float to decimal (significand, exponent, is_negative).

in from_chars, fast_float convert input chars into parsed_number_string, which has exponent, mantissa, negative.

I tried to write function from_decimal:

template<typename T>
std::errc from_decimal(int64_t exponent, uint64_t mantissa, bool negative, T& value) noexcept {

  static_assert (std::is_same<T, double>::value || std::is_same<T, float>::value, "only float and double are supported");

  parsed_number_string pns {.exponent = exponent, .mantissa = mantissa, .negative = negative, .valid = true};

and continue same as from_chars.
It is segfaulting, apparently you also need to fill integer and fraction in parsed_number_string. 
Looking at code, it is hard to figure out how correctly set integer and fraction in parsed_number_string, giving input of significand and exponent from Dragonbox.
Would you mind to give me a hint, how to calculate integer and fraction from significand and exponent?

Also... may be you can consider making from_decimal as member of public API? I am just touching this area, but sounds like converting decimal -> double can be useful.

Thank you for reading!

**Comment by lemire** [comment]
> Would you mind to give me a hint, how to calculate integer and fraction from significand and exponent?

What are you doing currently?

**Comment by scherepanov** [comment]
Currently I am converting float to double through intermediate chars presentation.
Definitely converting back and forth from/to chars is a performance issue.
Question - can I use decimal as intermediate format, avoiding chars.
That should be more efficient.

**Comment by lemire** [comment]
You definitely can pass parsed_number_string to the appropriate function. For obvious reasons, we are not going to support such uses. You are accessing internal functions and structures. It is not part of our public API. So we are not going to document or expose any of it. It may change or be removed in future releases.

We can still answer questions but it is expected that you have done the work. If you want to bypass the public interface, you will have to study the code and build your own components, at your own risks.

So as far as this issue is concerned, the answer is negative, we will not expose our internal functions to public usage. 

Feel free to ask questions, e.g., as part of the discussion section on GitHub. But we won’t provide documentation and support for this type of use. We may just answer specific questions.

**Comment by lemire** [comment]
> I am writing conversion float -> double, with requirement to_string(float) = to_string(double)

It is an intriguing problem and I would love to learn more about it.

Not at the expense of increasing the surface area and maintenance of our library, however.

**Comment by scherepanov** [comment]
Thank you for advice, will do as you suggested - study your code and figure out internal structures.

As you requested, here is a very real use case.

Users are generating a bunch of 4-bytes floats (in financial business), and it is loading into database, with intermediate CSV format. Database has only 8-bytes double datatype.
I got an idea to make process more efficient and use binary intermediate format for database loading, writing my own c++ parser that runs inside database. Flat values no longer converted to chars on client, and sent to my parser inside database as 4-bytes floats, where I am converting to double and load into table columns. With loading through binary format, users start seeing a bunch of random digits attached to float values in their SQL tools, and comparison in WHERE clause stop working the way it used to. That is a critical bug.
Generating of binary load format is around 7 times faster comparing to CSV. Loading from binary format used to be faster than loading from CSV, until I implemented float->double conversion, then it became slightly slower. 

I am trying to make my binary parser to be faster than CSV parser, by making float->double conversion more efficient.

Obvious improvement would be to avoid conversion back/from chars, and use decimal.

I put very descriptive story of my problem here https://github.com/jk-jeon/dragonbox/issues/28
